lib/matplotlib/pyplot.py

```diff
@@ -213,21 +216,61 @@ def findobj(o=None, match=None, include_self=True):
 
 def switch_backend(newbackend):
     """
-    Switch the default backend.  This feature is **experimental**, and
-    is only expected to work switching to an image backend.  e.g., if
-    you have a bunch of PostScript scripts that you want to run from
-    an interactive ipython session, you may want to switch to the PS
-    backend before running them to avoid having a bunch of GUI windows
-    popup.  If you try to interactively switch from one GUI backend to
-    another, you will explode.
+    Close all open figures and set the Matplotlib backend.
+
+    The argument is case-insensitive.  Switching to an interactive backend is
+    possible only if no event loop for another interactive backend has started.
+    Switching to and from non-interactive backends is always possible.
 
-    Calling this command will close all open windows.
+    Parameters
+    ----------
+    newbackend : str or List[str]
+        The name of the backend to use.  If a list of backends, they will be
+        tried in order until one successfully loads.
     """
     close('all')
+
+    if not isinstance(newbackend, str):
+        for candidate in newbackend:
+            try:
+                _log.info("Trying to load backend %s.", candidate)
+                return switch_backend(candidate)
+            except ImportError as exc:
+                _log.info("Loading backend %s failed: %s", candidate, exc)
+        else:
+            raise ValueError("No suitable backend among {}".format(newbackend))
+
+    backend_name = (
+        newbackend[9:] if newbackend.startswith("module://")
+        else "matplotlib.backends.backend_{}".format(newbackend.lower()))
+
+    backend_mod = importlib.import_module(backend_name)
+    Backend = type(
+        "Backend", (matplotlib.backends._Backend,), vars(backend_mod))
+    _log.info("Loaded backend %s version %s.",
+              newbackend, Backend.backend_version)
+
+    required_framework = Backend.required_interactive_framework
+    current_framework = \
+        matplotlib.backends._get_running_interactive_framework()
+    if (current_framework and required_framework
+            and current_framework != required_framework):
+        raise ImportError(
```

Comment:
I am worried that by this point we are already broken (due to importing the underlying GUI wrappers which may over-write each other's PyOS_EventHook), but I don't see any way around that short of delaying all of the imports (which I am not sure we _can_ do).

Reply:
PyOS_InputHook should (hopefully) not be overwritten before the toolkit event loop actually starts; e.g. for PyQt:
```
from PyQt5.QtWidgets import *
print(input("foo? "))
app = QApplication([])
print(input("foo? "))
```
even after the QApplication is created, `input` works fine; overwriting likely only occurs in `app.exec_()`.

Reply:
But if you the import tk will the Qt windows still be responsive while waiting for user input?